cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,722 @@
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             ArrowArray* out)
+{
+  using DeviceType = int32_t;
+  return decimals_to_arrow<DeviceType>(
+    std::move(column), cudf::detail::max_precision<DeviceType>(), stream, mr, out);
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<numeric::decimal64>(cudf::column&& column,
+                                                             rmm::cuda_stream_view stream,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             ArrowArray* out)
+{
+  using DeviceType = int64_t;
+  return decimals_to_arrow<DeviceType>(
+    std::move(column), cudf::detail::max_precision<DeviceType>(), stream, mr, out);
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<numeric::decimal128>(cudf::column&& column,
+                                                              rmm::cuda_stream_view stream,
+                                                              rmm::mr::device_memory_resource* mr,
+                                                              ArrowArray* out)
+{
+  using DeviceType = __int128_t;
+  return decimals_to_arrow<DeviceType>(
+    std::move(column), cudf::detail::max_precision<DeviceType>(), stream, mr, out);
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<bool>(cudf::column&& column,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr,
+                                               ArrowArray* out)
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), NANOARROW_TYPE_BOOL, column));
+
+  auto bitmask  = bools_to_mask(column.view(), stream, mr);
+  auto contents = column.release();
+  if (contents.null_mask) {
+    NANOARROW_RETURN_NOT_OK(
+      set_buffer(std::move(contents.null_mask), validity_buffer_idx, tmp.get()));
+  }
+  NANOARROW_RETURN_NOT_OK(
+    set_buffer(std::move(bitmask.first), fixed_width_data_buffer_idx, tmp.get()));
+
+  ArrowArrayMove(tmp.get(), out);
+  return NANOARROW_OK;
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::string_view>(cudf::column&& column,
+                                                            rmm::cuda_stream_view stream,
+                                                            rmm::mr::device_memory_resource* mr,
+                                                            ArrowArray* out)
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), NANOARROW_TYPE_STRING, column));
+
+  if (column.size() == 0) {
+    auto zero = std::make_unique<rmm::device_scalar<int32_t>>(0, stream, mr);
+    NANOARROW_RETURN_NOT_OK(set_buffer(std::move(zero), fixed_width_data_buffer_idx, tmp.get()));
```

Comment:
Is this strictly necessary? Quoting from the spec:
> The buffer pointers MAY be null only in two situations:
>
>    for the null bitmap buffer, if [ArrowArray.null_count](https://arrow.apache.org/docs/format/> CDataInterface.html#c.ArrowArray.null_count) is 0;
>
>    **for any buffer, if the size in bytes of the corresponding buffer would be 0.**

https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowArray.buffers

Reply:
That's not for the null bitmap, it's for the offsets buffer. Many implementations expect there to be n+1 offsets, so a 0-length string array needs to have a single 0 value for its offset buffer

Reply:
Sorry for any confusion, I had bolded the second point there, the one that says that you can leave the pointer null if the size of the corresponding buffer would be 0 for any buffer, not just the null mask. It sounds like from what you're saying that maybe many implementations are not allowing for that in this case, though, and they still dereference the offsets. Is that in violation of the spec, or am I reading it wrong?

Reply:
Ah interesting OK thanks for the clarifications. From our perspective it would certainly be nicer if we could avoid the unnecessary allocation, and I would have expected the spec to promise that implementations wouldn't dereference the offsets of an empty string column. Could you add a comment here explaining why we need to allocate this scalar?

Reply:
Added the comment

Reply:
Thanks!